### DISCUSSION TITLE
Stability of new DOM

### DISCUSSION THREAD

**Discussion by EfremBraun** [bokeh/bokeh#12082] [General]
I tried updating the version of Bokeh I was running from 3.0.0dev2 to the latest 3.0.0dev5, and I noticed that https://github.com/bokeh/bokeh/pull/11239 made some big changes to the DOM that broke a lot of my code (e.g., CSS selectors such as given in https://github.com/bokeh/bokeh/issues/11524#issuecomment-946081107 and those used in my Selenium tests).

I’m wondering how stable this change to the DOM is. It’ll take me a few man-hours to fix the issues, so I’d like to do it only once. If more changes are planned such that the final version version of 3.0 will be different from 3.0.0.dev5, I’ll hold off on making these updates for now, but if more isn’t planned, I’d prefer to fix things up now.

**Comment by EfremBraun** [comment]
@mattpap

**Comment by mattpap** [comment]
Redesign of the layout engine (PR #11915) will change the DOM-related internals of bokehjs quite a bit. Thus I would hold off upgrades until that work is finished. When the new layout is merged, we will be working with Holoviz team (Panel in particular) to stabilize CSS/DOM APIs and provide a migration path.

**Comment by EfremBraun** [comment]
Perhaps I'll add in a feature request for these changes, which is ensuring that they will still allow the user to make CSS tweaks to the internals of the Bokeh components.

As shown in https://github.com/bokeh/bokeh/issues/11524#issuecomment-946081107, I used to be able to do this by putting `custom_class` in the widget's `css_classes` argument, and then specifying the CSS style for elements of type `bk` with parent elements having class `custom_class`.

However, I don't think I can do this any longer with the Bokeh component internals being in a shadow tree, as per https://web.dev/shadowdom-v1/#styling-a-component-from-the-outside, "Users can tweak internal styles [only?] if the component's author provides styling hooks using [CSS custom properties](https://developer.mozilla.org/docs/Web/CSS/Using_CSS_variables)."

My application has a lot of this kind of styling, so at the moment, I think I'll be stuck without being able to make the style changes I'd like.

**Comment by EfremBraun** [comment]
Also, for what it's worth, per https://discourse.bokeh.org/t/including-css-classes-for-datatable/9055/5, it does seem like others Bokeh users are using the same trick I was for internal component styling.
